lean4-infoview/src/infoview/tooltips.tsx

```diff
@@ -1,73 +1,167 @@
+      }
+      return coords
+    }
+
+    const partial = (fn: (coords: Coords, side: Side) => Coords, side: Side) => {
+      return (coords: Coords) => { return fn(coords, side) }
+    }
+
+    type QuadTransforms = {
+      [k in string]: { [k: string]: (coords: Coords) => Coords }
+    }
+    const quadTransforms: QuadTransforms = {
+      'top-left': {'top': partial(flip, 'top'),},
+      'top-right': {'top': partial(flip, 'top'), 'right': partial(flip, 'right')},
+      'bottom-right': {'right': partial(flip, 'right')},
+      'bottom-left' : {},
+    }
+
+    // Split viewport into four quadrants.
+    type Quadrants = 'top-right' | 'top-left' | 'bottom-right' | 'bottom-left'
+    const getQuadrant = (x: number, y: number) : Quadrants => {
+      const vw = document.documentElement.clientWidth
+      const vh = document.documentElement.clientHeight
+      const hvw = vw / 2
+      const hvh = vh / 2
+
+      if (x <= hvw && y <= hvh) { return 'top-left' }
+      else if (x <= hvw && y > hvh) { return 'bottom-left' }
+      else if (x > hvw && y > hvh) { return 'bottom-right' }
+      else { return 'top-right' }
+    }
+
+    // Preferred position of tooltip is top-right relative to pointer.
+    let coords: Coords = {
+      x: pointerPos.pageX,
+      y: pointerPos.pageY - floatingRect.height - 10
+    }
+    state.x = coords.x
+    state.y = coords.y
+    let overflow : SideObject = await detectOverflow(state);
+    const quadrant = getQuadrant(pointerPos.clientX, pointerPos.clientY)
+
+    // Apply transformations if tooltip is overflowing. Transformations to
+    // be applied depend on where the pointer is located in the viewport.
+    if (containsPositive(overflow)) {
+      const transforms = quadTransforms[quadrant]
+      Object.entries(transforms).forEach(([side, transform]) => {
+        if (overflow[side as Side] > 1) { coords = transform(coords) }
+      })
+      state.x = coords.x
+      state.y = coords.y
+      overflow = await detectOverflow(state)
+
+      // If tooltip is still overflowing, make small adjustments.
+      if (containsPositive(overflow)) {
+        Object.entries(overflow)
+          .filter(([_, overflowAmount]) => overflowAmount > 1)
+          .forEach(([side, overflowAmount]) => {
+            switch(side as Side) {
+              case 'top': coords.y = document.documentElement.scrollTop + 1; break;
+              // case 'bottom': coords.y = coords.y + overflowAmount + 1; break;
+              case 'left': coords.x = coords.x + overflowAmount + 1; break;
+              case 'right': coords.x = coords.x - overflowAmount - 1; break;
+            }
+          })
+      }
+    }
+
+    // [subpixel accelerated positioning]
+    // (https://floating-ui.com/docs/misc#subpixel-and-accelerated-positioning).
+    Object.assign(floatingEl.style, {
+      top: '0',
+      left: '0',
+      transform: `translate(${roundByDPR(coords.x)}px,${roundByDPR(coords.y)}px)`,
+    });
```

Comment:
Rather than making an imperative update, could the `transform` be passed to the `div` returned from `Tooltip` directly? You can grab `x/y` from the return values of `useFloating`.